server/public/shared/request/context.go

```diff
@@ -84,72 +86,89 @@ func (c *Context) UserAgent() string {
 func (c *Context) AcceptLanguage() string {
 	return c.acceptLanguage
 }
-
+func (c *Context) Logger() mlog.LoggerIFace {
+	return c.logger
+}
 func (c *Context) Context() context.Context {
 	return c.context
 }
 
-func (c *Context) SetSession(s *model.Session) {
-	c.session = *s
-}
-
-func (c *Context) SetT(t i18n.TranslateFunc) {
-	c.t = t
+func (c *Context) WithT(t i18n.TranslateFunc) CTX {
+	rctx := c.clone()
+	rctx.t = t
+	return rctx
 }
-func (c *Context) SetRequestId(s string) {
-	c.requestId = s
+func (c *Context) WithSession(s *model.Session) CTX {
+	rctx := c.clone()
+	rctx.session = *s
+	return rctx
 }
-func (c *Context) SetIPAddress(s string) {
-	c.ipAddress = s
+func (c *Context) WithRequestId(s string) CTX {
+	rctx := c.clone()
+	rctx.requestId = s
+	return rctx
 }
-func (c *Context) SetXForwardedFor(s string) {
-	c.xForwardedFor = s
+func (c *Context) WithIPAddress(s string) CTX {
+	rctx := c.clone()
+	rctx.ipAddress = s
+	return rctx
 }
-func (c *Context) SetUserAgent(s string) {
-	c.userAgent = s
+func (c *Context) WithXForwardedFor(s string) CTX {
+	rctx := c.clone()
+	rctx.xForwardedFor = s
+	return rctx
 }
-func (c *Context) SetAcceptLanguage(s string) {
-	c.acceptLanguage = s
+func (c *Context) WithPath(s string) CTX {
+	rctx := c.clone()
+	rctx.path = s
+	return rctx
 }
-func (c *Context) SetPath(s string) {
-	c.path = s
+func (c *Context) WithUserAgent(s string) CTX {
+	rctx := c.clone()
+	rctx.userAgent = s
+	return rctx
 }
-func (c *Context) SetContext(ctx context.Context) {
-	c.context = ctx
+func (c *Context) WithAcceptLanguage(s string) CTX {
+	rctx := c.clone()
+	rctx.acceptLanguage = s
+	return rctx
 }
-
-func (c *Context) GetT() i18n.TranslateFunc {
-	return c.t
+func (c *Context) WithContext(ctx context.Context) CTX {
+	rctx := c.clone()
+	rctx.context = ctx
+	return rctx
 }
-
-func (c *Context) SetLogger(logger mlog.LoggerIFace) {
-	c.logger = logger
+func (c *Context) WithLogger(logger mlog.LoggerIFace) CTX {
+	rctx := c.clone()
+	rctx.logger = logger
+	return rctx
 }
 
-func (c *Context) Logger() mlog.LoggerIFace {
-	return c.logger
+func (c *Context) With(f func(ctx CTX) CTX) CTX {
+	return f(c)
 }
```

Comment:
Should this take care of the cloning as well? As in:
```go
func (c *Context) With(f func(ctx CTX) CTX) CTX {
	return f(c.clone())
}
```
It may be unexpected that all other `WithXYZ` functions do clone the context except for this generic one. Also, if it would clone it, we could actually use it everywhere else, as in:
```go
func (c * Context) WithUserAgent(s string) CTX {
	return c.With(func(c) { c.userAgent = s })
}
```

Reply:
If `With` takes care the cloning, it becomes less useful for any caller of `With`, as that would lead to cloning the context twice: Once within `With` and then as part of the called `WithX` method.

E.g. ` c.AppContext.With(sqlstore.RequestContextWithMaster)` would lead to one `clone` in `With` and one in `RequestContextWithMaster` via `WithContext`.